src/client/pythonEnvironments/common/commonUtils.ts

```
@@ -1,63 +1,94 @@
+        maxDepth: opts?.maxDepth,
+        ignoreErrors: opts.ignoreErrors === undefined ? false : opts.ignoreErrors,
+    };
+    return iterExecutables(root, 1, cfg);
+}
 
-    let dirContents: string[];
+export async function* iterExecutables(
+    root: string,
+    depth: number,
+    cfg: {
+        checkBin: FileFilterFunc;
+        filterFile: FileFilterFunc | undefined;
+        maxDepth: number | undefined;
+        ignoreErrors: boolean;
+    },
+): AsyncIterableIterator<string> {
+    let entries: Dirent[];
     try {
-        dirContents = (await listDir(root))
-            // Build the full filename.
-            .map((c) => path.join(root, c))
-            // Apply the filter.
-            .filter(itemFilter);
+        entries = await listDir(root);
     } catch (err) {
         // Treat a missing directory as empty.
         if (err.code === 'ENOENT') {
             return;
         }
+        if (cfg.ignoreErrors) {
+            logError(`listDir() failed for "${root}" (${err})`);
+            return;
+        }
         throw err; // re-throw
     }
 
-    const generators = dirContents.map((fullPath) => {
-        async function* generator() {
-            if (await isDirectory(fullPath)) {
-                if (recurseLevels && recurseLevels > 0) {
-                    yield* findInterpretersInDir(fullPath, recurseLevels - 1, filter);
+    for (const entry of entries) {
+        const filename = path.join(root, entry.name);
+        if (entry.isDirectory()) {
+            if (cfg.maxDepth && depth <= cfg.maxDepth) {
+                yield* iterExecutables(filename, depth + 1, cfg);
+            }
+        } else if (entry.isFile()) {
+            let matched = true;
+            if (cfg.filterFile) {
+                try {
+                    matched = cfg.filterFile(filename);
+                } catch (err) {
+                    if (cfg.ignoreErrors) {
+                        logError(`filterFile() failed for "${filename}" (${err})`);
```

I was expecting us to continue here instead of return.

---

Also, I expect to ignore errors and continue by default.